src/v/storage/disk_log_impl.cc

```diff
@@ -256,17 +256,27 @@ disk_log_impl::time_based_gc_max_offset(gc_config cfg) {
     // files so that size-based retention eventually evicts the problematic
     // segment, preventing a crash.
     //
+    // with Redpanda v23.3, segment_index tracks broker_timestamps, and it's the
+    // value returned by segment_index::retention_timestamp() for **new**
+    // segments. for older segments, the issue still remains and
+    // storage_ignore_timestamps_in_future_secs is another way to deal with it
+
+    // this will retrive cluster configs, to reused them during the scan since
+    // there are no scheduling points
```

Comment:
nit: maybe a bit extra, but it would be nice to ensure no scheduling points will be added here in the future. Could do this by using the `task_local` util Evgeny added.

```
task_local<bool> local_task_guard{true};
... protected section ...
vassert(local_task_guard.has_value(), "Oops - unexpected scheduling point");
```

Reply:
I checked it and it's a cool idea, but I didn't include it because the damage is small if a scheduling point is added and the relevant configs are updated, it means that they will take effect at the next round